src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -0,0 +1,283 @@
+import {
+  ExpectedVersionConflictError,
+  STREAM_DOES_NOT_EXIST,
+  assertExpectedVersionMatchesCurrent,
+  type EventStore,
+  type Event,
+  type AggregateStreamOptions,
+  type AggregateStreamResult,
+  type AppendToStreamOptions,
+  type AppendToStreamResult,
+  type ReadStreamOptions,
+  type ReadStreamResult,
+  type ReadEvent,
+  type ReadEventMetadata,
+  type ExpectedStreamVersion,
+} from '@event-driven-io/emmett';
+import { type Collection, MongoClient, type WithId } from 'mongodb';
+import { v4 as uuid } from 'uuid';
+
+export const MongoDBEventStoreDefaultStreamVersion = -1;
+export const MongoDBDefaultCollectionName = 'eventstreams';
+
+export type StreamType = string;
+export type StreamName<T extends StreamType = StreamType> = `${T}:${string}`;
+
+export type StreamNameParts<T extends StreamType = StreamType> = {
+  streamType: T;
+  entityId: string;
+};
+
+export type StreamToProject<EventType extends Event> = {
+  streamName: StreamName;
+  streamType: StreamType;
+  entityId: string;
+  streamVersion: number;
+  events: ReadEvent<EventType, ReadEventMetadata>[];
+};
+
+export interface EventStream<EventType extends Event> {
+  streamName: string;
+  events: Array<ReadEvent<EventType, ReadEventMetadata>>;
+  createdAt: Date;
+  updatedAt: Date;
+}
+export type EventStreamEvent<EventType extends Event> =
+  EventStream<EventType>['events'][number];
+
+export interface MongoDBConnectionOptions {
+  connectionString: string;
+  database: string;
+  collection?: string;
+}
+
+class EventStoreClass implements EventStore<number> {
+  private readonly collection: Collection<EventStream<Event>>;
+
+  constructor(collection: typeof this.collection) {
+    this.collection = collection;
+  }
+
+  async readStream<EventType extends Event>(
+    streamName: StreamName,
+    options?: ReadStreamOptions<number>,
+  ): Promise<Exclude<ReadStreamResult<EventType, number>, null>> {
+    const expectedStreamVersion = options?.expectedStreamVersion;
+
+    // @ts-expect-error
```

Comment:
The only two uses of `@ts-expect-error` here are because the `collection` has to have a specific type but the type of each event stream will differ depending on the specific stream.

This is just to satisfy that the stream returned has the expected TS types.

Reply:
We could probably get away with this by adjusting the `findOne` and `findOneAndUpdate` type defs to allow a generic argument for the specific type with some `declare module ...` magic, but I'm afraid how that will affect people using this library and also using `mongodb`.

Reply:
Technically we should define our collection type. The set of metadata is known, and the events array type can be just default Event without defaulted generic parameters.